internal/controller/engine/engine.go

```diff
@@ -0,0 +1,434 @@
+	}
+
+	e.mx.Lock()
+	e.controllers[name] = r
+	e.mx.Unlock()
+
+	return nil
+}
+
+// Stop a controller.
+func (e *ControllerEngine) Stop(ctx context.Context, name string) error {
+	e.mx.RLock()
+	c, running := e.controllers[name]
+	e.mx.RUnlock()
+
+	// Stop is a no-op if the controller isn't running.
+	if !running {
+		return nil
+	}
+
+	// First stop the controller's watches.
+	stopped, err := e.StopWatches(ctx, name)
+	if err != nil {
+		return errors.Wrapf(err, "cannot stop watches for controller %q", name)
+	}
+
+	// Then stop any informers that only this controller was using.
+	if stopped > 0 {
+		if err := e.RemoveUnwatchedInformers(ctx); err != nil {
+			return errors.Wrapf(err, "cannot remove unwatched informers after stopping watches for controller %q", name)
+		}
+	}
+
+	// Finally, stop and delete the controller.
+	e.mx.Lock()
+	c.cancel()
+	delete(e.controllers, name)
+	e.mx.Unlock()
+
+	e.log.Debug("Stopped controller", "controller", name)
+	return nil
+}
+
+// IsRunning returns true if the named controller is running.
+func (e *ControllerEngine) IsRunning(name string) bool {
+	e.mx.RLock()
+	_, running := e.controllers[name]
+	e.mx.RUnlock()
+	return running
+}
+
+// Watch an object.
+type Watch struct {
+	kind       client.Object
+	handler    handler.EventHandler
+	predicates []predicate.Predicate
+}
+
+// WatchFor returns a Watch for the supplied kind of object. Events will be
+// handled by the supplied EventHandler, and may be filtered by the supplied
+// predicates.
+func WatchFor(kind client.Object, h handler.EventHandler, p ...predicate.Predicate) Watch {
+	return Watch{kind: kind, handler: h, predicates: p}
+}
+
+// StartWatches instructs the named controller to start the supplied watches.
+// The controller will only start a watch if it's not already watching the type
+// of object specified by the supplied Watch. StartWatches blocks other
+// operations on the same controller if and when it starts a watch.
+func (e *ControllerEngine) StartWatches(name string, ws ...Watch) error {
+	e.mx.RLock()
+	c, ok := e.controllers[name]
+	e.mx.RUnlock()
+
+	if !ok {
+		return errors.Errorf("controller %q is not running", name)
+	}
+
+	watchExists := make(map[schema.GroupVersionKind]bool)
+	c.mx.RLock()
+	for gvk := range c.sources {
+		watchExists[gvk] = true
+	}
+	c.mx.RUnlock()
+
+	// It's possible that we didn't explicitly stop a watch, but its backing
+	// informer was removed. This implicitly stops the watch by deleting its
+	// backing listener. If a watch exists but doesn't have an active informer,
+	// we want to restart the watch (and, implicitly, the informer).
+	a := e.infs.ActiveInformers()
+	activeInformer := make(map[schema.GroupVersionKind]bool, len(a))
+	for _, gvk := range a {
+		activeInformer[gvk] = true
+	}
+
+	// Using a map here deduplicates watches by GVK. If we're asked to start
+	// several watches for the same GVK at in the same call, we'll only start
+	// the last one.
```

Comment:
is this what you want? Watches can have different predicates or even a different handler.

Reply:
Maybe not. 🤔 

I had considered this, but I couldn't think of a case where one controller would want to watch the same GVK twice with different handlers/predicates.

Now that I think on it some more, there is one (edge?) case though - an XR could be composed of the same kind of XR. If an XR was composed of the same kind of XR I'd expect:

* The controller would start watching the XR GVK with `EnqueueRequestForObject{}`
* The controller would therefore never start watching the XR GVK with `EnqueueCompositeResources`

So if XR A composes XR B, we'd wouldn't a reconcile for XR A when B changed.

Reply:
I pushed a commit that addresses this by adding a `WatchType`, which callers can use to distinguish multiple watches for the same GVK for different reasons. Two watches for the same GVK with different handlers or predicates should use different `WatchTypes`.

The only alternative that came to mind was using reflection to get the handler/predicate types, but making folks explicitly distinguish the types feels less fragile and magic.